# 1. Preparation

Import libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


Import dataframe from task b

In [24]:
toronto = pd.read_csv('boroughs_location.csv', index_col=0)
print(toronto.shape)
toronto.head()

(102, 5)


PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Harbourfront, Regent Park  43.654260   
3        M6A        North York  Lawrence Heights, Lawrence Manor  43.718518   
4        M7A      Queen's Park                      Queen's Park  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

to ease the analytical task, limit to the Boroughs that contain the name "Toronto"

In [25]:
toronto = toronto[toronto['Borough'].str.contains("Toronto")].reset_index(drop=True)

In [26]:
print(toronto.shape)
toronto.head()

(38, 5)


PostalCode           Borough               Neighborhood   Latitude  \
0        M5A  Downtown Toronto  Harbourfront, Regent Park  43.654260   
1        M5B  Downtown Toronto   Ryerson, Garden District  43.657162   
2        M5C  Downtown Toronto             St. James Town  43.651494   
3        M4E      East Toronto                The Beaches  43.676357   
4        M5E  Downtown Toronto                Berczy Park  43.644771   

   Longitude  
0 -79.360636  
1 -79.378937  
2 -79.375418  
3 -79.293031  
4 -79.373306

38 Boroughs containg Toronto.
Now we get Toronto's coordinates and display the map with the neighborhoods

In [5]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.653963, -79.387207.


In [27]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto['Latitude'], toronto['Longitude'], 
                                           toronto['Borough'],toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_toronto) 
    
map_toronto

# 2. get nearby venues from Foursquare

next cell is supposed to be hidden

In [7]:
# @hidden_cell
CLIENT_ID = '4HVWRS4XLAGRNWY2LALOCOGC330FP4JMYJSLIX5KLJI5HI43' # your Foursquare ID
CLIENT_SECRET = 'GLPWFEFJT3OMRHR2T43B5KWEFYWLDYCEMK4O4ISAYSAHIX2P' # your Foursquare Secret
VERSION = '20190211' # Foursquare API version
LIMIT = 100
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 4HVWRS4XLAGRNWY2LALOCOGC330FP4JMYJSLIX5KLJI5HI43
CLIENT_SECRET:GLPWFEFJT3OMRHR2T43B5KWEFYWLDYCEMK4O4ISAYSAHIX2P


In [28]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [29]:
toronto_venues = getNearbyVenues(names=toronto['Neighborhood'],
                                   latitudes=toronto['Latitude'],
                                   longitudes=toronto['Longitude']
                                  )

Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Adelaide, King, Richmond
Dovercourt Village, Dufferin
Harbourfront East, Toronto Islands, Union Station
Little Portugal, Trinity
The Danforth West, Riverdale
Design Exchange, Toronto Dominion Centre
Brockton, Exhibition Place, Parkdale Village
The Beaches West, India Bazaar
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North, Forest Hill West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
Harbord, University of Toronto
Runnymede, Swansea
Moore Park, Summerhill East
Chinatown, Grange Park, Kensington Market
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Rosedale
Stn A PO Boxes 25 The Esplanade
Cabbagetown, St. James Town
Fir

In [30]:
print(toronto_venues.shape)
toronto_venues.head()

(1708, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Harbourfront, Regent Park               43.65426              -79.360636   
1  Harbourfront, Regent Park               43.65426              -79.360636   
2  Harbourfront, Regent Park               43.65426              -79.360636   
3  Harbourfront, Regent Park               43.65426              -79.360636   
4  Harbourfront, Regent Park               43.65426              -79.360636   

                                             Venue  Venue Latitude  \
0                                 Roselle Desserts       43.653447   
1                                    Tandem Coffee       43.653559   
2  Toronto Cooper Koo Family Cherry St YMCA Centre       43.653191   
3                               Morning Glory Cafe       43.653947   
4                              Body Blitz Spa East       43.654735   

   Venue Longitude        Venue Category  
0       -79.362017                Bakery  
1       -79.361809           Coffee Shop  
2       -79.357947  Gym / Fitness Center  
3       -79.361149        Breakfast Spot  
4       -79.359874                   Spa

We got 1,708 venues. Now let's see how many venues in each Borough

In [31]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            56   
Brockton, Exhibition Place, Parkdale Village                           18   
Business Reply Mail Processing Centre 969 Eastern                      18   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            48   
Central Bay Street                                                     82   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               16   
Church and Wellesley                                                   87   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             39   
Davisville North                                                       11   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     15   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           19   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     4   
Harbord, University of Toronto                                         35   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              47   
High Park, The Junction South                                          23   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               64   
Moore Park, Summerhill East                                             4   
North Toronto West                                                     21   
Parkdale, Roncesvalles                                                 16   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     39   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        94   
Studio District                                                        39   
The Annex, North Midtown, Yorkville                                    25   
The Beaches                                                             4   
The Beaches West, India Bazaar                                         19   
The Danforth West, Riverdale                                           42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             56   
Brockton, Exhibition Place, Parkdale Village                            18   
Business Reply Mail Processing Centre 969 Eastern                       18   
CN Tower, Bathurst Quay, Island airport, Harbou...                      14   
Cabbagetown, St. James Town                                             48   
Central Bay Street                                                      82   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                16   
Church and Wellesley                                                    87   
Commerce Court, Victoria Hotel     

In [32]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 242 unique categories.


In [65]:
toronto

PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M5B  Downtown Toronto   
2         M5C  Downtown Toronto   
3         M4E      East Toronto   
4         M5E  Downtown Toronto   
5         M5G  Downtown Toronto   
6         M6G  Downtown Toronto   
7         M5H  Downtown Toronto   
8         M6H      West Toronto   
9         M5J  Downtown Toronto   
10        M6J      West Toronto   
11        M4K      East Toronto   
12        M5K  Downtown Toronto   
13        M6K      West Toronto   
14        M4L      East Toronto   
15        M5L  Downtown Toronto   
16        M4M      East Toronto   
17        M4N   Central Toronto   
18        M5N   Central Toronto   
19        M4P   Central Toronto   
20        M5P   Central Toronto   
21        M6P      West Toronto   
22        M4R   Central Toronto   
23        M5R   Central Toronto   
24        M6R      West Toronto   
25        M4S   Central Toronto   
26        M5S  Downtown Toronto   
27        M6S      West Toronto   
28        M4T   Central Toronto   
29        M5T  Downtown Toronto   
30        M4V   Central Toronto   
31        M5V  Downtown Toronto   
32        M4W  Downtown Toronto   
33        M5W  Downtown Toronto   
34        M4X  Downtown Toronto   
35        M5X  Downtown Toronto   
36        M4Y  Downtown Toronto   
37        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                           Harbourfront, Regent Park  43.654260 -79.360636  
1                            Ryerson, Garden District  43.657162 -79.378937  
2                                      St. James Town  43.651494 -79.375418  
3                                         The Beaches  43.676357 -79.293031  
4                                         Berczy Park  43.644771 -79.373306  
5                                  Central Bay Street  43.657952 -79.387383  
6                                            Christie  43.669542 -79.422564  
7                            Adelaide, King, Richmond  43.650571 -79.384568  
8                        Dovercourt Village, Dufferin  43.669005 -79.442259  
9   Harbourfront East, Toronto Islands, Union Station  43.640816 -79.381752  
10                           Little Portugal, Trinity  43.647927 -79.419750  
11                       The Danforth West, Riverdale  43.679557 -79.352188  
12           Design Exchange, Toronto Dominion Centre  43.647177 -79.381576  
13       Brockton, Exhibition Place, Parkdale Village  43.636847 -79.428191  
14                     The Beaches West, India Bazaar  43.668999 -79.315572  
15                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
16                                    Studio District  43.659526 -79.340923  
17                                      Lawrence Park  43.728020 -79.388790  
18                                           Roselawn  43.711695 -79.416936  
19                                   Davisville North  43.712751 -79.390197  
20                Forest Hill North, Forest Hill West  43.696948 -79.411307  
21                      High Park, The Junction South  43.661608 -79.464763  
22                                 North Toronto West  43.715383 -79.405678  
23                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
24                             Parkdale, Roncesvalles  43.648960 -79.456325  
25                                         Davisville  43.704324 -79.388790  
26                     Harbord, University of Toronto  43.662696 -79.400049  
27                                 Runnymede, Swansea  43.651571 -79.484450  
28                        Moore Park, Summerhill East  43.689574 -79.383160  
29          Chinatown, Grange Park, Kensington Market  43.653206 -79.400049  
30  Deer Park, Forest Hill SE, Rathnelly, South Hi...  43.686412 -79.400049  
31  CN Tower, Bathurst Quay, Island airport, Harbo...  43.628947 -79.394420  
32                                           Rosedale  43.67956

# 3. Analyse neighborhoods

In [33]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Adult Boutique  Afghan Restaurant  Airport  \
0            0               0                  0        0   
1            0               0                  0        0   
2            0               0                  0        0   
3            0               0                  0        0   
4            0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                 0                    0             0         0   
1                 0                    0             0         0   
2                 0                    0             0         0   
3                 0                    0             0         0   
4                 0                    0             0         0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                 0            0           0                    0   
1                 0            0           0                    0   
2                 0            0           0                    0   
3                 0            0           0                    0   
4                 0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       1     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Beer Bar  Beer Store  Belgian Restaurant  Bistro  Board Shop  \
0         0           0                   0       0           0   
1         0           0                   0       0           0   
2         0           0                   0       0           0   
3         0           0                   0       0           0   
4         0           0                   0       0           0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               1   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaur

In [34]:
toronto_onehot.shape

(1708, 242)

1,708 venues in 242 categories.

In [35]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                         Berczy Park     0.000000   
2        Brockton, Exhibition Place, Parkdale Village     0.000000   
3   Business Reply Mail Processing Centre 969 Eastern     0.055556   
4   CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
5                         Cabbagetown, St. James Town     0.000000   
6                                  Central Bay Street     0.012195   
7           Chinatown, Grange Park, Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.011494   
10                     Commerce Court, Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000   
14           Design Exchange, Toronto Dominion Centre     0.000000   
15                       Dovercourt Village, Dufferin     0.000000   
16             First Canadian Place, Underground city     0.000000   
17                Forest Hill North, Forest Hill West     0.000000   
18                     Harbord, University of Toronto     0.000000   
19  Harbourfront East, Toronto Islands, Union Station     0.000000   
20                          Harbourfront, Regent Park     0.000000   
21                      High Park, The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                           Little Portugal, Trinity     0.015625   
24                        Moore Park, Summerhill East     0.000000   
25                                 North Toronto West     0.047619   
26                             Parkdale, Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                 Runnymede, Swansea     0.000000   
30                           Ryerson, Garden District     0.000000   
31                                     St. James Town     0.010000   
32                    Stn A PO Boxes 25 The Esplanade     0.010638   
33                                    Studio District     0.025641   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                     The Beaches West, India Bazaar     0.000000   
37                       The Danforth West, Riverdale     0.023810   

    Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0         0.000000           0.000000  0.000000            0.000000   
1         0.000000           0.000000  0.000000            0.000000   
2         0.000000           0.000000  0.000000            0.000000   
3         0.000000           0.000000  0.000000            0.000000   
4         0.000000           0.000000  0.071429            0.071429   
5         0.000000           0.000000  0.000000            0.000000   
6         0.000000           0.000000  0.000000            0.000000   
7         0.000000           0.000000  0.000000            0.000000   
8         0.000000           0.000000  0.000000            0.000000   
9         0.011494           0.011494  0.000000            0.000000   
10        0.000000           0.000000  0.000000            0.000000   
11        0.000000           0.000000  0.000000            0.000000   
12        0.000000           0.000000  0.000000            0.000000   
13        0.000000           0.000000  0.000000            0.000000   
14        0.000000           0.000000  0.000000            0.000000   
15        0.000000           0.000000  0.000000            0.000000   
16        0.000000           0.000000  0.000000            0.000000   
17        0.000000           0.0

In [36]:
toronto_grouped.shape

(38, 242)

In [46]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [68]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(
                                                 toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café   American Restaurant            Steakhouse   
1            Restaurant          Cocktail Bar           Cheese Shop   
2        Breakfast Spot                  Café             Pet Store   
3           Yoga Studio         Auto Workshop         Garden Center   
4      Airport Terminal       Airport Service              Boutique   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Thai Restaurant        Clothing Store      Asian Restaurant   
1                  Café                   Pub        Farmers Market   
2         Burrito Place               Stadium                   Bar   
3                Garden  Fast Food Restaurant        Farmers Market   
4      Sculpture Garden         Boat or Ferry                 Plane   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant                 Hotel                    Bar  
1            Steakhouse    Italian Restaurant                 Bakery  
2  Caribbean Restaurant    Italian Restaurant           Climbing Gym  
3                  Park            Comic Shop       Recording Studio  
4          Airport Gate    Airport Food Court                Airport

# 4.Cluster Neighborhoods

We cluster the neighborhoods and 

In [69]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [70]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.tail(6) # check the last columns!

PostalCode           Borough  \
32        M4W  Downtown Toronto   
33        M5W  Downtown Toronto   
34        M4X  Downtown Toronto   
35        M5X  Downtown Toronto   
36        M4Y  Downtown Toronto   
37        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
32                                           Rosedale  43.679563 -79.377529   
33                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
34                        Cabbagetown, St. James Town  43.667967 -79.367675   
35             First Canadian Place, Underground city  43.648429 -79.382280   
36                               Church and Wellesley  43.665860 -79.383160   
37  Business Reply Mail Processing Centre 969 Eastern  43.662744 -79.321558   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
32               1                  Park            Playground   
33               0           Coffee Shop            Restaurant   
34               0           Coffee Shop            Restaurant   
35               0                  Café           Coffee Shop   
36               0   Japanese Restaurant      Sushi Restaurant   
37               0    Light Rail Station           Yoga Studio   

   3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
32                 Trail  Eastern European Restaurant        Discount Store   
33                  Café                     Beer Bar    Italian Restaurant   
34           Pizza Place                       Bakery    Italian Restaurant   
35                 Hotel                   Restaurant   American Restaurant   
36           Coffee Shop                      Gay Bar          Burger Joint   
37         Auto Workshop                Garden Center                Garden   

   6th Most Common Venue 7th Most Common Venue     8th Most Common Venue  \
32               Dog Run      Doner Restaurant                Donut Shop   
33    Seafood Restaurant                   Pub                     Hotel   
34                  Café                   Pub                    Market   
35         Deli / Bodega    Seafood Restaurant                       Bar   
36            Restaurant  Fast Food Restaurant  Mediterranean Restaurant   
37  Fast Food Restaurant        Farmers Market                      Park   

   9th Most Common Venue 10th Most Common Venue  
32   Dumpling Restaurant          Women's Store  
33          Cocktail Bar                 Bakery  
34             Pet Store               Pharmacy  
35            Steakhouse       Asian Restaurant  
36                  Café            Men's Store  
37            Comic Shop       Recording Studio

In [71]:
neighborhoods_venues_sorted.head()

Cluster Labels                                       Neighborhood  \
0               0                           Adelaide, King, Richmond   
1               0                                        Berczy Park   
2               0       Brockton, Exhibition Place, Parkdale Village   
3               0  Business Reply Mail Processing Centre 969 Eastern   
4               0  CN Tower, Bathurst Quay, Island airport, Harbo...   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop                  Café   American Restaurant   
1           Coffee Shop            Restaurant          Cocktail Bar   
2           Coffee Shop        Breakfast Spot                  Café   
3    Light Rail Station           Yoga Studio         Auto Workshop   
4        Airport Lounge      Airport Terminal       Airport Service   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0            Steakhouse       Thai Restaurant        Clothing Store   
1           Cheese Shop                  Café                   Pub   
2             Pet Store         Burrito Place               Stadium   
3         Garden Center                Garden  Fast Food Restaurant   
4              Boutique      Sculpture Garden         Boat or Ferry   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0      Asian Restaurant            Restaurant                 Hotel   
1        Farmers Market            Steakhouse    Italian Restaurant   
2                   Bar  Caribbean Restaurant    Italian Restaurant   
3        Farmers Market                  Park            Comic Shop   
4                 Plane          Airport Gate    Airport Food Court   

  10th Most Common Venue  
0                    Bar  
1                 Bakery  
2           Climbing Gym  
3       Recording Studio  
4                Airport

Now we create the map shwing the clusters of Toronto Neighborhoods

In [72]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], 
                                  toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster+1), parse_html=True)
    cluster = int(cluster)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],  
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
    
map_clusters

Examining the Clusters

### Cluster 1

In this cluster we find mainly coffe food shops and entertainment options. This cluster contains the vast majority of Boroughs.

In [58]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0        Clothing Store   
2   Downtown Toronto               0           Coffee Shop   
3       East Toronto               0           Coffee Shop   
4   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0                  Café   
7   Downtown Toronto               0           Coffee Shop   
8       West Toronto               0           Supermarket   
9   Downtown Toronto               0           Coffee Shop   
10      West Toronto               0                   Bar   
11      East Toronto               0      Greek Restaurant   
12  Downtown Toronto               0           Coffee Shop   
13      West Toronto               0           Coffee Shop   
14      East Toronto               0                  Park   
15  Downtown Toronto               0           Coffee Shop   
16      East Toronto               0                  Café   
19   Central Toronto               0      Asian Restaurant   
21      West Toronto               0                  Café   
22   Central Toronto               0   Sporting Goods Shop   
23   Central Toronto               0           Coffee Shop   
24      West Toronto               0        Breakfast Spot   
25   Central Toronto               0           Pizza Place   
26  Downtown Toronto               0                  Café   
27      West Toronto               0           Coffee Shop   
29  Downtown Toronto               0                  Café   
30   Central Toronto               0           Coffee Shop   
31  Downtown Toronto               0        Airport Lounge   
33  Downtown Toronto               0           Coffee Shop   
34  Downtown Toronto               0           Coffee Shop   
35  Downtown Toronto               0                  Café   
36  Downtown Toronto               0   Japanese Restaurant   
37      East Toronto               0    Light Rail Station   

   2nd Most Common Venue  3rd Most Common Venue  \
0                   Café                 Bakery   
1            Coffee Shop                   Café   
2             Restaurant                   Café   
3                    Pub           Burger Joint   
4             Restaurant           Cocktail Bar   
5                   Café     Italian Restaurant   
6          Grocery Store                   Park   
7                   Café    American Restaurant   
8               Pharmacy         Discount Store   
9                  Hotel               Aquarium   
10           Men's Store             Restaurant   
11           Coffee Shop         Ice Cream Shop   
12                  Café                  Hotel   
13        Breakfast Spot                   Café   
14        Ice Cream Shop                Brewery   
15                  Café             Restaurant   
16           Coffee Shop                 Bakery   
19          Burger Joint   Gym / Fitness Center   
21    Mexican Restaurant            Flea Market   
22           Coffee Shop         Clothing Store   
23                  Café         Sandwich Place   
24             Gift Shop              Bookstore   
25          Dessert Shop         Sandwich Place   
26           Coffee Shop                 Bakery   
27                  Café            Pizza Place   
29                   Bar  Vietnamese Restaurant   
30                   Pub            Pizza Place   
31      Airport Terminal        Airport Service   
33            Restaurant                   Café   
34            Restaurant            Pizza Place   
35           Coffee Shop                  Hotel   
36      Sushi Restaurant            Coffee Shop   
37           Yoga Studio          Auto Workshop   

            4th Most Common Venue      5th Most Common Venue  \
0                            Park                        Pub   
1                  Cosmetics Shop  Middle

### Cluster 2

This cluster is dominated by parks, playground and trails

In [59]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
20   Central Toronto               1         Jewelry Store   
32  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
20      Sushi Restaurant                  Park                        Trail   
32            Playground                 Trail  Eastern European Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
20         Women's Store   Dumpling Restaurant               Dog Run   
32        Discount Store               Dog Run      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
20      Doner Restaurant            Donut Shop      Electronics Store  
32            Donut Shop   Dumpling Restaurant          Women's Store

### Cluster 3

This one Borough is dominated by gardens

In [60]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               2                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18         Women's Store     Fish & Chips Shop  Fast Food Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18        Farmers Market    Falafel Restaurant           Event Space   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
18  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant

### Cluster 4

This cluster is dominated by transportation and restaurants

In [63]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
17  Central Toronto               3              Bus Line   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17    Dim Sum Restaurant           Swim School                  Park   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17  Eastern European Restaurant               Dog Run      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17            Donut Shop   Dumpling Restaurant          Women's Store

### Cluster 5

A kid's frinedly neighborhood with sport options

In [66]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
28  Central Toronto               4            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
28                   Gym                 Trail          Tennis Court   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
28         Women's Store   Dumpling Restaurant        Discount Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
28               Dog Run      Doner Restaurant             Donut Shop